compiler-wrapper not resolved from upstream install_tree when using separate user install_tree

### Steps to reproduce

We upgraded our central installation to Spack 1.1.1 and introduced a new install tree:
```
  install_tree:
    root: /cluster/spack/opt/rub_2026
```
The users receive an `upstreams.yaml`:
```
upstreams:
  central-spack:
    install_tree: /cluster/spack/opt/rub_2026
```
and have a local configuration:
```
config:
  install_tree:
    root: $HOME/spack/opt/spack
```
Without installing anything locally, running:
```
[esseramj@login004 ~]$ spack location -i compiler-wrapper
```
returns
```
/home/esseramj/spack/opt/spack/linux-zen4/compiler-wrapper-1.0-cusdudx5rjxfvarlsljjiiq7celgz3fq
```
However, for any other package (example: xz):
```
[esseramj@login004 ~]$ spack location -i xz
```
correctly returns the upstream location:
```
/cluster/spack/opt/rub_2026/linux-zen4/xz-5.6.3-izlubxa5dr637hzq4jox7fai3ml5dv5j`
```
This happens only for compiler-wrapper.

With debug outputs:
```
[esseramj@login004 ~]$ spack -d location -i xz
==> [2026-02-19-16:40:32.353174] Reading config from file /cluster/spack/1.1.1/etc/spack/defaults/base/config.yaml
==> [2026-02-19-16:40:32.366232] Attempting to read from missing scope: <ConfigScope: system> at /etc/spack
==> [2026-02-19-16:40:32.366313] Reading config from file /cluster/spack/1.1.1/etc/spack/site/config.yaml
==> [2026-02-19-16:40:32.385172] Reading config from file /home/esseramj/.spack/config.yaml
==> [2026-02-19-16:40:32.393063] Imported location from built-in commands
==> [2026-02-19-16:40:32.393457] Imported location from built-in commands
==> [2026-02-19-16:40:32.393751] Attempting to read from missing scope: <ConfigScope: system> at /etc/spack
==> [2026-02-19-16:40:32.393834] Reading config from file /cluster/spack/1.1.1/etc/spack/toolchains.yaml
==> [2026-02-19-16:40:32.398362] Attempting to read from missing scope: <ConfigScope: system> at /etc/spack
==> [2026-02-19-16:40:32.398515] Reading config from file /home/esseramj/.spack/upstreams.yaml
==> [2026-02-19-16:40:32.399501] DATABASE LOCK TIMEOUT: 120s
==> [2026-02-19-16:40:32.470175] DATABASE LOCK TIMEOUT: 60s
==> [2026-02-19-16:40:32.470263] PACKAGE LOCK TIMEOUT: No timeout
/cluster/spack/opt/rub_2026/linux-zen4/xz-5.6.3-izlubxa5dr637hzq4jox7fai3ml5dv5j
```
and
```
[esseramj@login004 ~]$ spack -d location -i compiler-wrapper
==> [2026-02-19-16:40:51.684905] Reading config from file /cluster/spack/1.1.1/etc/spack/defaults/base/config.yaml
==> [2026-02-19-16:40:51.697889] Attempting to read from missing scope: <ConfigScope: system> at /etc/spack
==> [2026-02-19-16:40:51.697976] Reading config from file /cluster/spack/1.1.1/etc/spack/site/config.yaml
==> [2026-02-19-16:40:51.716765] Reading config from file /home/esseramj/.spack/config.yaml
==> [2026-02-19-16:40:51.723393] Imported location from built-in commands
==> [2026-02-19-16:40:51.723796] Imported location from built-in commands
==> [2026-02-19-16:40:51.724084] Attempting to read from missing scope: <ConfigScope: system> at /etc/spack
==> [2026-02-19-16:40:51.724165] Reading config from file /cluster/spack/1.1.1/etc/spack/toolchains.yaml
==> [2026-02-19-16:40:51.728634] Attempting to read from missing scope: <ConfigScope: system> at /etc/spack
==> [2026-02-19-16:40:51.728709] Reading config from file /home/esseramj/.spack/upstreams.yaml
==> [2026-02-19-16:40:51.729703] DATABASE LOCK TIMEOUT: 120s
==> [2026-02-19-16:40:51.799568] DATABASE LOCK TIMEOUT: 60s
==> [2026-02-19-16:40:51.799663] PACKAGE LOCK TIMEOUT: No timeout
==> [2026-02-19-16:40:51.800979] Reading config from file /cluster/spack/1.1.1/etc/spack/defaults/base/packages.yaml
==> [2026-02-19-16:40:51.821976] Attempting to read from missing scope: <ConfigScope: system> at /etc/spack
==> [2026-02-19-16:40:51.822203] Reading config from file /cluster/spack/1.1.1/etc/spack/packages.yaml
==> [2026-02-19-16:40:51.831964] Reading config from file /cluster/spack/1.1.1/etc/spack/defaults/base/repos.yaml
==> [2026-02-19-16:40:51.833476] Attempting to read from missing scope: <ConfigScope: system> at /etc/spack
==> [2026-02-19-16:40:51.833710] Reading config from file /cluster/spack/1.1.1/etc/spack/repos.yaml
/home/esseramj/spack/opt/spack/linux-zen4/compiler-wrapper-1.0-cusdudx5rjxfvarlsljjiiq7celgz3fq
```

Since the `compiler-wrapper` is a dependency of ´gcc`, no packages can be installed as a user:
```
==> Installing openmpi-5.0.8-tsrv2afcygipb57t2njxzap6w42qxul7 [41/41]
==> Fetching https://mirror.spack.io/_source-cache/archive/53/53131e1a57e7270f645707f8b0b65ba56048f5b5ac3f68faabed3eb0d710e449.tar.bz2
    [100%]   30.29 MB @  173.4 MB/s
==> No patches needed for openmpi
==> openmpi: Executing phase: 'autoreconf'
==> openmpi: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 77:
    '/tmp/esseramj/spack-stage/spack-stage-openmpi-5.0.8-tsrv2afcygipb57t2njxzap6w42qxul7/spack-src/configure' '--prefix=/home/esseramj/spack/opt/spack/linux-zen4/openmpi-5.0.8-tsrv2afcygipb57t2njxzap6w42qxul7' '--enable-shared' '--disable-silent-rules' '--disable-sphinx' '--enable-builtin-atomics' '--disable-static' '--enable-mpi1-compatibility' '--without-psm2' '--without-cma' '--without-psm' '--without-xpmem' '--without-verbs' '--without-hcoll' '--without-fca' '--with-ofi=/cluster/spack/opt/rub_2026/linux-zen4/libfabric-2.4.0-use65iwwsqfv5tlhdn3qa76umukx3khz' '--without-ucx' '--without-mxm' '--without-ucc' '--without-knem' '--without-xpmem' '--without-alps' '--with-slurm' '--without-tm' '--without-sge' '--without-lsf' '--without-loadleveler' '--disable-memchecker' '--with-libevent=/cluster/spack/opt/rub_2026/linux-zen4/libevent-2.1.12-vki7qkgf3zw5u6n7mg44rizguhopwbe3' '--with-pmix=/cluster/spack/opt/rub_2026/linux-zen4/pmix-6.0.0-bvd3lwfiznoz7prf5kxh6g6yezbtmuxj' '--with-prrte=/cluster/spack/opt/rub_2026/linux-zen4/prrte-4.0.0-w4gxmlg4thidk3qip6ylfe227flsngzy' '--with-zlib=/cluster/spack/opt/rub_2026/linux-zen4/zlib-ng-2.3.2-ytl5qpripldqs54swa3amfnerkffk5y2' '--with-hwloc=/cluster/spack/opt/rub_2026/linux-zen4/hwloc-2.12.2-ug3lar6zz4olsatxf4fwaxj3vzbujqlw' '--disable-java' '--disable-mpi-java' '--disable-io-romio' '--with-gpfs=no' '--without-cuda' '--without-rocm' '--enable-wrapper-rpath' '--disable-wrapper-runpath' '--enable-mpi-fortran' 'CFLAGS=-DYY_BUF_SIZE=1048576' '--disable-debug'

2 errors found in build log:
     125    ============================================================================
     126    
     127    *** C compiler and preprocessor
     128    checking whether make supports the include directive... yes (GNU style)
     129    checking for gcc... /home/esseramj/spack/opt/spack/linux-zen4/compiler-wrapper-1.0-cusdudx5rjxfvarlsljjiiq7celgz3fq/libexec/spack/gcc/gcc
     130    checking whether the C compiler works... no
  >> 131    configure: error: in `/tmp/esseramj/spack-stage/spack-stage-openmpi-5.0.8-tsrv2afcygipb57t2njxzap6w42qxul7/spack-src':
  >> 132    configure: error: C compiler cannot create executables
     133    See `config.log' for more details

See build log for details:
  /tmp/esseramj/spack-stage/spack-stage-openmpi-5.0.8-tsrv2afcygipb57t2njxzap6w42qxul7/spack-build-out.txt
```

As admin, it resolves correctly to the central tree.
```
[hpcxko3d@login004 ~]$ spack location -i compiler-wrapper
/cluster/spack/opt/rub_2026/linux-zen4/compiler-wrapper-1.0-cusdudx5rjxfvarlsljjiiq7celgz3fq
```

As local user I have tried using `spack gc`, `spack clean -a` and `rm -rf /home/$USER/spack/opt/spack/.spack-db/` but it still resolves to the local path.

I don't think it's a local artifact since I cannot uninstall the package (of course, it is installed in upstream):
```
[esseramj@login004 ~]$ spack uninstall /cusdud
==> Error: /cusdud does not match any installed packages.
```

### Error message

_No response_

### Information on your system

```
* **Spack:** 1.1.1
* **Builtin repo:** https://github.com/spack/spack-packages/commit/dd6b1cede2c91e9d9d250677aa94b9f37b492c3b
* **Python:** 3.9.25
* **Platform:** linux-almalinux9-zen4
```

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output

--- a/lib/spack/spack/new_installer.py
+++ b/lib/spack/spack/new_installer.py
@@ -1725,8 +1725,7 @@ class ScheduleResult(NamedTuple):
 def schedule_builds(
     pending: List[str],
     build_graph: BuildGraph,
-    db: spack.database.Database,
-    prefix_locker: spack.database.SpecLocker,
+    store: spack.store.Store,
     overwrite: Set[str],
     overwrite_time: float,
     capacity: int,
@@ -1749,8 +1748,7 @@ def schedule_builds(
     Args:
         pending: List of dag hashes pending installation; modified in-place.
         build_graph: The build dependency graph; used for node lookup and parent enqueueing.
-        db: Package database; used for read lock and installed-status queries.
-        prefix_locker: Per-spec write locker.
+        store: The local store for installs.
         overwrite: Set of dag hashes to overwrite even if already installed.
         overwrite_time: Timestamp (from time.time()) at which the overwrite install was requested.
             A spec in ``overwrite`` whose DB installation_time >= overwrite_time was installed by
@@ -1768,6 +1766,7 @@ def schedule_builds(
     newly_installed: List[Tuple[str, spack.spec.Spec, spack.util.lock.Lock]] = []
     to_mark_explicit: List[MarkExplicitAction] = []
     blocked = True
+    db = store.db
 
     # Acquire the DB read lock non-blocking; hold it throughout the loop so the in-memory snapshot
     # stays consistent while we acquire per-spec prefix locks.
@@ -1779,7 +1778,7 @@ def schedule_builds(
         while capacity and idx < len(pending):
             dag_hash = pending[idx]
             spec = build_graph.nodes[dag_hash]
-            lock = prefix_locker.lock(spec)
+            lock = store.prefix_locker.lock(spec)
 
             if lock.try_acquire_write():
                 blocked = False
@@ -2503,8 +2502,7 @@ class PackageInstaller:
         result = schedule_builds(
             pending=self.pending_builds,
             build_graph=self.build_graph,
-            db=self.db,
-            prefix_locker=spack.store.STORE.prefix_locker,
+            store=spack.store.STORE,  # todo: dependency injection
             overwrite=self.overwrite,
             overwrite_time=self.overwrite_time,
             capacity=self.capacity,

--- a/lib/spack/spack/new_installer.py
+++ b/lib/spack/spack/new_installer.py
@@ -1819,12 +1819,16 @@ def schedule_builds(
                 continue
 
             # Write lock acquired: proceed with scheduling.
-            # Don't schedule builds for specs from upstream databases.
-            if upstream and record and not record.installed:
-                lock.release_write()
-                raise spack.error.InstallError(
-                    f"Cannot install {spec}: it is uninstalled in an upstream database."
-                )
+
+            # Allow local installs of specs that are referenced but uninstalled in an upstream db.
+            # In edge cases where an upstream has forcibly uninstalled just a link dep, we cannot
+            # fix the situation by installing it locally, because dependents in the upstream
+            # reference the missing upstream path. However, we document that users should not
+            # remove installed specs from upstreams, so it's a relatively safe assumption that
+            # missing upstream specs were never installed or garbage collected build deps, and are
+            # not referenced by absolute path in installed dependents.
+            if upstream and record and not record.installed and not spec.external:
+                spec.set_prefix(store.layout.path_for_spec(spec))
 
             # Defensively assert prefix invariants
             if not spec.external:

--- a/lib/spack/spack/test/conftest.py
+++ b/lib/spack/spack/test/conftest.py
@@ -4,6 +4,7 @@
 
 import base64
 import collections
+import contextlib
 import datetime
 import email.message
 import errno
@@ -36,6 +37,7 @@ import spack.compilers.config
 import spack.compilers.libraries
 import spack.concretize
 import spack.config
+import spack.database
 import spack.directives_meta
 import spack.environment as ev
 import spack.error
@@ -1351,6 +1353,55 @@ def gen_mock_layout(tmp_path: Path):
         return MockLayout(str(subroot))
 
     yield create_layout
+
+
+@contextlib.contextmanager
+def writable(database):
+    """Allow a database to be written inside this context manager."""
+    old_lock, old_is_upstream = database.lock, database.is_upstream
+    db_root = Path(database.root)
+
+    try:
+        # this is safe on all platforms during tests (tests get their own tmpdirs)
+        database.lock = spack.util.lock.Lock(str(database._lock_path), enable=False)
+        database.is_upstream = False
+        db_root.chmod(mode=0o755)
+        with database.write_transaction():
+            yield
+    finally:
+        db_root.chmod(mode=0o555)
+        database.lock = old_lock
+        database.is_upstream = old_is_upstream
+
+
+@pytest.fixture()
+def upstream_and_downstream_db(tmp_path: Path, gen_mock_layout):
+    """Fixture for a pair of stores: upstream and downstream.
+
+    Upstream API prohibits writing to an upstream, so we also return a writable version
+    of the upstream DB for tests to use.
+
+    """
+    mock_db_root = tmp_path / "mock_db_root"
+    mock_db_root.mkdir()
+    mock_db_root.chmod(0o555)
+
+    upstream_db = spack.database.Database(
+        str(mock_db_root), is_upstream=True, layout=gen_mock_layout("a")
+    )
+    with writable(upstream_db):
+        upstream_db._write()
+
+    downstream_db_root = tmp_path / "mock_downstream_db_root"
+    downstream_db_root.mkdir()
+    downstream_db_root.chmod(0o755)
+
+    downstream_db = spack.database.Database(
+        str(downstream_db_root), upstream_dbs=[upstream_db], layout=gen_mock_layout("b")
+    )
+    downstream_db._write()
+
+    yield upstream_db, downstream_db
 
 
 class ConfigUpdate:

--- a/lib/spack/spack/test/database.py
+++ b/lib/spack/spack/test/database.py
@@ -3,7 +3,6 @@
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 """Check the database is functioning properly, both in memory and in its file."""
 
-import contextlib
 import datetime
 import functools
 import gzip
@@ -44,59 +43,10 @@ from spack.enums import InstallRecordStatus
 from spack.installer import PackageInstaller
 from spack.llnl.util.tty.colify import colify
 from spack.schema.database_index import schema
-from spack.test.conftest import RepoBuilder
+from spack.test.conftest import RepoBuilder, writable
 from spack.util.executable import Executable
 
 pytestmark = pytest.mark.db
-
-
-@contextlib.contextmanager
-def writable(database):
-    """Allow a database to be written inside this context manager."""
-    old_lock, old_is_upstream = database.lock, database.is_upstream
-    db_root = pathlib.Path(database.root)
-
-    try:
-        # this is safe on all platforms during tests (tests get their own tmpdirs)
-        database.lock = lk.Lock(str(database._lock_path), enable=False)
-        database.is_upstream = False
-        db_root.chmod(mode=0o755)
-        with database.write_transaction():
-            yield
-    finally:
-        db_root.chmod(mode=0o555)
-        database.lock = old_lock
-        database.is_upstream = old_is_upstream
-
-
-@pytest.fixture()
-def upstream_and_downstream_db(tmp_path: pathlib.Path, gen_mock_layout):
-    """Fixture for a pair of stores: upstream and downstream.
-
-    Upstream API prohibits writing to an upstream, so we also return a writable version
-    of the upstream DB for tests to use.
-
-    """
-    mock_db_root = tmp_path / "mock_db_root"
-    mock_db_root.mkdir()
-    mock_db_root.chmod(0o555)
-
-    upstream_db = spack.database.Database(
-        str(mock_db_root), is_upstream=True, layout=gen_mock_layout("a")
-    )
-    with writable(upstream_db):
-        upstream_db._write()
-
-    downstream_db_root = tmp_path / "mock_downstream_db_root"
-    downstream_db_root.mkdir()
-    downstream_db_root.chmod(0o755)
-
-    downstream_db = spack.database.Database(
-        str(downstream_db_root), upstream_dbs=[upstream_db], layout=gen_mock_layout("b")
-    )
-    downstream_db._write()
-
-    yield upstream_db, downstream_db
 
 
 @pytest.mark.parametrize(

--- a/lib/spack/spack/test/new_installer.py
+++ b/lib/spack/spack/test/new_installer.py
@@ -12,7 +12,12 @@ import pytest
 if sys.platform == "win32":
     pytest.skip("No Windows support", allow_module_level=True)
 
+from typing import Tuple
+
+import spack.deptypes as dt
 import spack.spec
+import spack.store
+from spack.database import Database
 from spack.new_installer import (
     OVERWRITE_GARBAGE_SUFFIX,
     BinaryCacheMiss,
@@ -22,7 +27,9 @@ from spack.new_installer import (
     _node_to_roots,
     schedule_builds,
 )
+from spack.new_installer_base import NoopJobServer
 from spack.new_installer_posix import PosixJobServer
+from spack.test.conftest import writable
 from spack.test.traverse import create_dag
 
 
@@ -627,6 +634,53 @@ class TestScheduleBuilds:
             for _, _, lock in result.newly_installed:
                 lock.release_read()
             jobserver.close()
+
+    def test_missing_in_upstream_is_installed_locally(
+        self, upstream_and_downstream_db: Tuple[Database, Database], mock_packages
+    ):
+        """A spec that is referenced but not installed in an upstream database is scheduled for
+        a local build, with its prefix repointed to the local store."""
+        upstream_db, downstream_db = upstream_and_downstream_db
+        dep = self._make_spec("dependency-install")
+        parent = spack.spec.Spec("dependent-install")
+        parent._add_dependency(dep, depflag=dt.BUILD, virtuals=())
+        parent._mark_concrete()
+
+        # Register both specs in the upstream, then uninstall the dep: its record is kept as
+        # uninstalled because the parent still references it.
+        with writable(upstream_db):
+            upstream_db.add(parent, explicit=True)
+            upstream_db.remove(dep)
+
+        # Create a Store for schedule_builds based on the downstream database.
+        local_store = spack.store.Store(downstream_db.root, upstreams=[upstream_db])
+        upstream, record = local_store.db.query_by_spec_hash(dep.dag_hash())
+        assert upstream and record is not None and not record.installed
+
+        # Like BuildGraph, start out with the prefix from the upstream record.
+        assert upstream_db.layout
+        dep.set_prefix(upstream_db.layout.path_for_spec(dep))
+
+        jobserver = NoopJobServer(num_jobs=2)
+        result = schedule_builds(
+            pending=[dep.dag_hash()],
+            build_graph=_FakeBuildGraph([dep]),  # type: ignore[arg-type]
+            store=local_store,
+            overwrite=set(),
+            overwrite_time=0.0,
+            capacity=1,
+            needs_jobserver_token=False,
+            jobserver=jobserver,
+            explicit=set(),
+        )
+        assert not result.blocked
+        assert [dag_hash for dag_hash, _ in result.to_start] == [dep.dag_hash()]
+        # The prefix was repointed from the upstream to the local store.
+        assert dep.prefix == local_store.layout.path_for_spec(dep)
+
+        # Cleanup
+        for _, lock in result.to_start:
+            lock.release_write()
 
 
 def test_nodes_to_roots():

--- a/lib/spack/spack/test/new_installer.py
+++ b/lib/spack/spack/test/new_installer.py
@@ -334,8 +334,7 @@ class TestScheduleBuilds:
             result = schedule_builds(
                 pending,
                 bg,
-                temporary_store.db,
-                temporary_store.prefix_locker,
+                temporary_store,
                 overwrite=set(),
                 overwrite_time=0.0,
                 capacity=1,
@@ -364,8 +363,7 @@ class TestScheduleBuilds:
             result = schedule_builds(
                 pending,
                 bg,
-                temporary_store.db,
-                temporary_store.prefix_locker,
+                temporary_store,
                 overwrite=set(),
                 overwrite_time=0.0,
                 capacity=1,
@@ -394,8 +392,7 @@ class TestScheduleBuilds:
             result = schedule_builds(
                 pending,
                 bg,
-                temporary_store.db,
-                temporary_store.prefix_locker,
+                temporary_store,
                 overwrite=set(),
                 overwrite_time=0.0,
                 capacity=2,
@@ -424,8 +421,7 @@ class TestScheduleBuilds:
             result = schedule_builds(
                 pending,
                 bg,
-                temporary_store.db,
-                temporary_store.prefix_locker,
+                temporary_store,
                 overwrite=set(),
                 overwrite_time=0.0,
                 capacity=2,
@@ -451,8 +447,7 @@ class TestScheduleBuilds:
             result = schedule_builds(
                 pending,
                 bg,
-                temporary_store.db,
-                temporary_store.prefix_locker,
+                temporary_store,
                 overwrite={spec.dag_hash()},
                 overwrite_time=time.time() + 100,
                 capacity=1,
@@ -484,8 +479,7 @@ class TestScheduleBuilds:
             result = schedule_builds(
                 pending,
                 bg,
-                temporary_store.db,
-                temporary_store.prefix_locker,
+                temporary_store,
                 overwrite=set(),
                 overwrite_time=0.0,
                 capacity=2,
@@ -523,8 +517,7 @@ class TestScheduleBuilds:
             result = schedule_builds(
                 pending,
                 bg,
-                temporary_store.db,
-                temporary_store.prefix_locker,
+                temporary_store,
                 overwrite=set(),
                 overwrite_time=0.0,
                 capacity=2,
@@ -562,8 +555,7 @@ class TestScheduleBuilds:
             result = schedule_builds(
                 pending,
                 bg,
-                temporary_store.db,
-                temporary_store.prefix_locker,
+                temporary_store,
                 overwrite=set(),
                 overwrite_time=0.0,
                 capacity=2,
@@ -589,8 +581,7 @@ class TestScheduleBuilds:
             result = schedule_builds(
                 pending,
                 bg,
-                temporary_store.db,
-                temporary_store.prefix_locker,
+                temporary_store,
                 overwrite={spec.dag_hash()},
                 overwrite_time=0.0,  # earlier than now()
                 capacity=1,
@@ -621,8 +612,7 @@ class TestScheduleBuilds:
             result = schedule_builds(
                 pending,
                 bg,
-                temporary_store.db,
-                temporary_store.prefix_locker,
+                temporary_store,
                 overwrite=set(),
                 overwrite_time=0.0,
                 capacity=1,

